cpp/unreal_plugins/SpCore/Source/SpCore/Unreal.h

```diff
@@ -212,40 +212,61 @@ class SPCORE_API Unreal
     static TReturnAsSceneComponent* createComponentInsideOwnerConstructor(UObject* owner, USceneComponent* parent, const std::string& scene_component_name)
+    {
+        return static_cast<TReturnAsSceneComponent*>(createComponentInsideOwnerConstructorByClass(TSceneComponent::StaticClass(), owner, parent, scene_component_name));
+    }
+
+    template <CSceneComponent TSceneComponent, CSceneComponent TReturnAsSceneComponent = TSceneComponent> requires
+        std::derived_from<TSceneComponent, TReturnAsSceneComponent>
+    static TReturnAsSceneComponent* createComponentInsideOwnerConstructor(USceneComponent* owner, const std::string& scene_component_name)
+    {
+        return createComponentInsideOwnerConstructor<TSceneComponent, TReturnAsSceneComponent>(owner, owner, scene_component_name);
+    }
+
+    static UObject* createComponentInsideOwnerConstructorByClass(UClass* component_class, AActor* owner, const std::string& component_name)
+    {
+        SP_ASSERT(owner);
+        UObject* object = owner->CreateDefaultSubobject(toFName(component_name), component_class, component_class, true, false);
+        SP_ASSERT(object);
+        return object;
+    }
+
+    static USceneComponent* createComponentInsideOwnerConstructorByClass(UClass* component_class, UObject* owner, USceneComponent* parent, const std::string& scene_component_name)
     {
         SP_ASSERT(owner);
         SP_ASSERT(parent);
-        TReturnAsSceneComponent* scene_component = owner->CreateDefaultSubobject<TSceneComponent>(toFName(scene_component_name));
+        USceneComponent* scene_component = static_cast<USceneComponent*>(owner->CreateDefaultSubobject(toFName(scene_component_name), component_class, component_class, true, false));
         SP_ASSERT(scene_component);
         scene_component->SetupAttachment(parent);
         return scene_component;
     }
 
-    template <CSceneComponent TSceneComponent, CSceneComponent TReturnAsSceneComponent = TSceneComponent> requires
-        std::derived_from<TSceneComponent, TReturnAsSceneComponent>
-    static TReturnAsSceneComponent* createComponentInsideOwnerConstructor(USceneComponent* owner, const std::string& scene_component_name)
+    static UObject* createComponentInsideOwnerConstructorByClassAsRoot(UClass* component_class, AActor* owner, const std::string& scene_component_name)
     {
-        return createComponentInsideOwnerConstructor<TSceneComponent, TReturnAsSceneComponent>(owner, owner, scene_component_name);
+        SP_ASSERT(owner);
+
+        USceneComponent* scene_component = static_cast<USceneComponent*>(owner->CreateDefaultSubobject(toFName(scene_component_name), component_class, component_class, true, false));
+        SP_ASSERT(scene_component);
+        owner->SetRootComponent(scene_component);
+        return scene_component;
+    }
+
+    static UObject* createComponentInsideOwnerConstructorByClass(UClass* component_class, USceneComponent* owner, const std::string& scene_component_name)
```

Comment:
Rename: `createComponentInsideOwnerConstructorByClass`
To: `createSceneComponentInsideOwnerConstructorByClass`

Also see feedback above.